### PR TITLE
ci: Only write docker build images to Cirrus cache

### PR DESCRIPTION
**Pull request by maflcko**
The `DOCKER_BUILD_CACHE_ARG` env var holds the options on how to use cache providers. Storing the image layers is useful for the Cirrus cache provider, because it offers 10GB per runner (https://cirrus-runners.app/setup/#speeding-up-the-cache). The cached image layers can help to avoid issues when the upstream package manager infra (apt native, apt llvm, pip, apk, git clone, ...) has outages or network issues.

However, on the GitHub Actions cache provider, a *total* cache of 10GB is offered for the whole repo. This cache must be shared with the depends cache, and the ccache, as well as the previous releases cache. So it is already full and trying to put the docker build layers into it will lead to an overflow.

Fix it by only writing to the docker cache on Cirrus.

Also, `DOCKER_BUILD_CACHE_ARG` requires a `shellcheck disable=SC2086` on the full build command. Fix that as well by using `shlex.split` from Python on just this variable.

### CODE REVIEW

File: ci/test/02_run_container.py [hunk 1]
```diff
@@ -45,6 +58,30 @@ def main():
                 file.write(f"{k}={v}\n")
     run(["cat", env_file])
 
+    if not os.getenv("DANGER_RUN_CI_ON_HOST"):
+        CI_IMAGE_LABEL = "bitcoin-ci-test"
+
+        maybe_cache_arg = shlex.split(os.getenv('DOCKER_BUILD_CACHE_ARG'))
```
Comment (cedwies) [inline ci/test/02_run_container.py]: What if we run this outside CI (where env var might not exist)? In that case `os.getenv(...)` returns `None` and `shlex.split(None)` throws `TypeError`.

Would
```py
maybe_cache_arg = shlex.split(os.getenv('DOCKER_BUILD_CACHE_ARG', ''))
```
make sense here? CI behavior is identical, but if it's missing locally we just get an empty list (no cache flags) instead of crashing. Or would you rather keep CI-only usability?

File: ci/test/02_run_container.py [hunk 2]
```diff
@@ -17,6 +18,18 @@ def run(cmd, **kwargs):
...
 
+def maybe_cpuset():
+    # Env vars during the build can not be changed. For example, a modified
+    # $MAKEJOBS is ignored in the build process. Use --cpuset-cpus as an
+    # approximation to respect $MAKEJOBS somewhat, if cpuset is available.
+    if os.getenv("HAVE_CGROUP_CPUSET"):
+        P = int(run(['nproc'], stdout=subprocess.PIPE).stdout)
+        M = min(P, int(os.getenv("MAKEJOBS").lstrip("-j")))
+        cpus = sorted(random.sample(range(P), M))
+        return [f"--cpuset-cpus={','.join(map(str, cpus))}"]
```
Comment (cedwies) [inline ci/test/02_run_container.py]: we compute `M` from `MAKEJOBS` and then always emit a cpuset.
CI normally uses `-jN` with N > 0, but for edge/local setups, if `MAKEJOBS` ends up as `-j0` (or missing/malformed?), `M` can be `0` -> cpus is `[]` → we emit `--cpuset-cpus= (empty)`. Would it be ok to skip the flag when M <= 0 (and optionally guard parsing)?

```py
def maybe_cpuset():
    if os.getenv("HAVE_CGROUP_CPUSET"):
        P = int(run(['nproc'], stdout=subprocess.PIPE).stdout)
        mj = os.getenv("MAKEJOBS", "-j1")  # maybe a default for local runs?
        try:
            M = min(P, int(mj.lstrip("-j")))
        except (TypeError, ValueError):
            return []  # bad/missing MAKEJOBS -> just skip?
        if M > 0:
            cpus = sorted(random.sample(range(P), M))
            return [f"--cpuset-cpus={','.join(map(str, cpus))}"]
    return []
```

This does not change normal CI behavior but just avoids an odd empty cpuset in edge cases. If you prefer strictness, we could alternatively `sys.exit` instead?

File: ci/test/02_run_container.sh [hunk 3]
```diff
@@ -10,30 +10,6 @@ export CI_IMAGE_LABEL="bitcoin-ci-test"
...
-      --build-arg "CI_IMAGE_NAME_TAG=${CI_IMAGE_NAME_TAG}" \
-      --build-arg "FILE_ENV=${FILE_ENV}" \
-      --build-arg "BASE_ROOT_DIR=${BASE_ROOT_DIR}" \
-      $MAYBE_CPUSET \
-      --platform="${CI_IMAGE_PLATFORM}" \
-      --label="${CI_IMAGE_LABEL}" \
-      --tag="${CONTAINER_NAME}" \
-      $DOCKER_BUILD_CACHE_ARG \
-      "${BASE_READ_ONLY_DIR}"
-
```
Comment (cedwies) [inline ci/test/02_run_container.sh:36]: Maybe a short comment is helpful for future maintainers?
`# The Docker build step has moved into 02_run_container.py (see PR #33639)`

**Review by cedwies — COMMENTED** [review]
Approach ACK

File: ci/test/02_run_container.sh (hunk 3 above)
Comment (maflcko) [reply ci/test/02_run_container.sh:36]: I think it is fine without comment, because the files are numbered in order and not too many, so one can just read/grep them all to search what they are looking for.

File: ci/test/02_run_container.py (hunk 2 above)
Comment (maflcko) [reply ci/test/02_run_container.py]: I don't think `MAKEJOBS=-j0` was ever supported by the CI. It will eventually be passed down to the test runner, which fails:

```
test/functional/test_runner.py", line 714, in __init__
    assert num_tests_parallel >= 1
           ^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

So I think I'll leave this as-is for now, to keep the code move-only, as the goal is not to change it. It can be changed in a separate PR, if there is need.

Just as a hint: If you want to use `nproc`, with the current CI system, you can just omit all settings, because it is picked by default:

https://github.com/bitcoin/bitcoin/blob/e744fd1249bf9577274614eaf3997bf4bbb612ff/ci/test/00_setup_env.sh#L37-L38

File: ci/test/02_run_container.py (hunk 1 above)
Comment (maflcko) [reply ci/test/02_run_container.py]: Ah, very nice catch. I had considered this case, but forgot to set the empty string before pushing the branch. Fixed now and pushed.

**Review by cedwies — COMMENTED** [review]
utACK fa29ecd
Reviewed the code and the new CI flow. The change to limit cache writes to Cirrus (while still reading cache on GHA) and moving the build step into Python with `shlex.split(os.getenv("DOCKER_BUILD_CACHE_ARG", ""))` looks good to me. I don’t see any unintended behavior changes beyond what is described in the PR.

File: ci/test/02_run_container.py [hunk 4]
```diff
@@ -17,6 +18,18 @@ def run(cmd, **kwargs):
...
 
 
+def maybe_cpuset():
+    # Env vars during the build can not be changed. For example, a modified
+    # $MAKEJOBS is ignored in the build process. Use --cpuset-cpus as an
+    # approximation to respect $MAKEJOBS somewhat, if cpuset is available.
+    if os.getenv("HAVE_CGROUP_CPUSET"):
+        P = int(run(["nproc"], stdout=subprocess.PIPE).stdout)
+        M = min(P, int(os.getenv("MAKEJOBS").lstrip("-j")))
+        cpus = sorted(random.sample(range(P), M))
```
Comment (l0rinc) [inline ci/test/02_run_container.py]: hmmm, not sure I get why we need this randomness here, is it to allow multiple runs without synchronization?

File: ci/test/02_run_container.py [hunk 5]
```diff
@@ -17,6 +18,18 @@ def run(cmd, **kwargs):
         sys.exit(e)
 
 
+def maybe_cpuset():
+    # Env vars during the build can not be changed. For example, a modified
```
Comment (l0rinc) [inline ci/test/02_run_container.py]: Can't we use Docker ARG to achieve the same result, or is that something different? The comment mentions env vars can't be changed during build, but ARG can be set at build time.

File: ci/test/02_run_container.py [hunk 6]
```diff
@@ -45,6 +58,30 @@ def main():
...
+        cmd_build += [
+            f"--file={os.getenv('BASE_READ_ONLY_DIR')}/ci/test_imagefile",
+            f"--build-arg=CI_IMAGE_NAME_TAG={os.getenv('CI_IMAGE_NAME_TAG')}",
+            f"--build-arg=FILE_ENV={os.getenv('FILE_ENV')}",
+            f"--build-arg=BASE_ROOT_DIR={os.getenv('BASE_ROOT_DIR')}",
+            f"--platform={os.getenv('CI_IMAGE_PLATFORM')}",
+            f"--label={CI_IMAGE_LABEL}",
+            f"--tag={os.getenv('CONTAINER_NAME')}",
+        ]
+        cmd_build += maybe_cpuset()
```
Comment (l0rinc) [inline ci/test/02_run_container.py]: I assume that changing the `MAYBE_CPUSET` position is irrelevant, right?

File: ci/test/02_run_container.py [hunk 7]
```diff
@@ -45,6 +58,30 @@ def main():
...
+        cmd_build += maybe_cache_arg
```
Comment (l0rinc) [inline ci/test/02_run_container.py]: for symmetry can we inline + unpack this?

File: ci/test/02_run_container.py [hunk 8]
```diff
@@ -45,6 +58,30 @@ def main():
...
+        cmd_build += [os.getenv("BASE_READ_ONLY_DIR")]
+
+        print(f"Building {os.getenv('CONTAINER_NAME')} image tag to run in")
```
Comment (l0rinc) [inline ci/test/02_run_container.py]: nit: we were logging `CI_IMAGE_NAME_TAG` before (just resolve if that's intentional)

File: ci/test/02_run_container.py [hunk 9]
```diff
@@ -45,6 +58,30 @@ def main():
...
+        run(cmd_build)
```
Comment (l0rinc) [inline ci/test/02_run_container.py:67]: Can we inline `cmd_build`?
```python
run(["docker", "buildx", "build",
    f"--file={os.environ['BASE_READ_ONLY_DIR']}/ci/test_imagefile",
    f"--build-arg=CI_IMAGE_NAME_TAG={os.environ['CI_IMAGE_NAME_TAG']}",
    f"--build-arg=FILE_ENV={os.environ['FILE_ENV']}",
    f"--build-arg=BASE_ROOT_DIR={os.environ['BASE_ROOT_DIR']}",
    f"--platform={os.environ['CI_IMAGE_PLATFORM']}",
    f"--label={CI_IMAGE_LABEL}",
    f"--tag={os.environ['CONTAINER_NAME']}",
    *maybe_cpuset(),
    *shlex.split(os.getenv("DOCKER_BUILD_CACHE_ARG", "")),
    os.environ["BASE_READ_ONLY_DIR"]
])
```

File: ci/test/02_run_container.py [hunk 10]
```diff
@@ -17,6 +18,18 @@ def run(cmd, **kwargs):
         sys.exit(e)
 
 
+def maybe_cpuset():
+    # Env vars during the build can not be changed. For example, a modified
+    # $MAKEJOBS is ignored in the build process. Use --cpuset-cpus as an
+    # approximation to respect $MAKEJOBS somewhat, if cpuset is available.
+    if os.getenv("HAVE_CGROUP_CPUSET"):
```
Comment (l0rinc) [inline ci/test/02_run_container.py]: Do we still need this? My understanding is that we're not really using this - can we delete and add back when there's a well-defined usecase for it?

If we do, we might want to add `MAKEJOBS` guard as well since that is also needed

File: .github/actions/configure-docker/action.yml [hunk 11]
```diff
@@ -48,8 +48,9 @@ runs:
         # Always optimistically --cache‑from in case a cache blob exists
         args=(--cache-from "type=gha${url_args:+,${url_args}},scope=${CONTAINER_NAME}")
 
-        # If this is a push to the default branch, also add --cache‑to to save the cache
-        if [[ ${{ github.event_name }} == "push" && ${{ github.ref_name }} == ${{ github.event.repository.default_branch }} ]]; then
+        # Only when using the Cirrus cache provider and this is a push to the default
+        # branch, add --cache-to to save the cache.
+        if [[ ${{ inputs.cache-provider }} == 'cirrus' && ${{ github.event_name }} == "push" && ${{ github.ref_name }} == ${{ github.event.repository.default_branch }} ]]; then
```
Comment (l0rinc) [inline .github/actions/configure-docker/action.yml:52]: slightly tangential, but maybe there's room for optimizing the docker layers by putting the variable parts as the last layers, using ARGs over ENVs, combining multiple layers, pinning versions and paths and variables early, pruning old layers, multi-stage builds, etc.

File: ci/test/02_run_container.py [hunk 12]
```diff
@@ -45,6 +58,30 @@ def main():
                 file.write(f"{k}={v}\n")
     run(["cat", env_file])
 
+    if not os.getenv("DANGER_RUN_CI_ON_HOST"):
+        CI_IMAGE_LABEL = "bitcoin-ci-test"
+
+        maybe_cache_arg = shlex.split(os.getenv("DOCKER_BUILD_CACHE_ARG", ""))
```
Comment (l0rinc) [inline ci/test/02_run_container.py]: Is this related to the mentioned https://www.shellcheck.net/wiki/SC2086? Because we need split to handle quotes and escaping properly?

File: ci/test/02_run_container.py [hunk 13]
```diff
@@ -17,6 +18,18 @@ def run(cmd, **kwargs):
...
 
+def maybe_cpuset():
+    # Env vars during the build can not be changed. For example, a modified
+    # $MAKEJOBS is ignored in the build process. Use --cpuset-cpus as an
+    # approximation to respect $MAKEJOBS somewhat, if cpuset is available.
+    if os.getenv("HAVE_CGROUP_CPUSET"):
+        P = int(run(["nproc"], stdout=subprocess.PIPE).stdout)
+        M = min(P, int(os.getenv("MAKEJOBS").lstrip("-j")))
+        cpus = sorted(random.sample(range(P), M))
+        return [f"--cpuset-cpus={','.join(map(str, cpus))}"]
```
Comment (l0rinc) [inline ci/test/02_run_container.py]: I don't yet understand why we're pinning CPUs here, what problem is that solving that `--cpus=` wouldn't?
The comment is a bit misleading since my understanding is that `--cpuset-cpus` doesn't actually limit CPU usage.

File: .github/actions/configure-docker/action.yml [hunk 14]
```diff
@@ -48,8 +48,9 @@ runs:
         # Always optimistically --cache‑from in case a cache blob exists
         args=(--cache-from "type=gha${url_args:+,${url_args}},scope=${CONTAINER_NAME}")
 
-        # If this is a push to the default branch, also add --cache‑to to save the cache
-        if [[ ${{ github.event_name }} == "push" && ${{ github.ref_name }} == ${{ github.event.repository.default_branch }} ]]; then
+        # Only when using the Cirrus cache provider and this is a push to the default
+        # branch, add --cache-to to save the cache.
```
Comment (l0rinc) [inline .github/actions/configure-docker/action.yml]:
```suggestion
        # Only add --cache-to when using the Cirrus cache provider and pushing to the default branch.
```

File: ci/test/02_run_container.py [hunk 15]
```diff
@@ -45,6 +58,30 @@ def main():
...
+    if not os.getenv("DANGER_RUN_CI_ON_HOST"):
+        CI_IMAGE_LABEL = "bitcoin-ci-test"
+
+        maybe_cache_arg = shlex.split(os.getenv("DOCKER_BUILD_CACHE_ARG", ""))
+
+        # Use buildx unconditionally
+        # Using buildx is required to properly load the correct driver, for use with registry caching. Neither build, nor BUILDKIT=1 currently do this properly
+        cmd_build = ["docker", "buildx", "build"]
+        cmd_build += [
+            f"--file={os.getenv('BASE_READ_ONLY_DIR')}/ci/test_imagefile",
```
Comment (l0rinc) [inline ci/test/02_run_container.py]: Doing `os.environ['BASE_READ_ONLY_DIR']` would rather raise a `KeyError` and the failure pinpoints the source of failure better

**Review by l0rinc — APPROVED** [review]
Concept ACK, I understand that you already have an ACK but please consider my suggestions

File: ci/test/02_run_container.py (hunk 15 above)
Comment (maflcko) [reply ci/test/02_run_container.py]: Ah nice, changed all

File: .github/actions/configure-docker/action.yml (hunk 14 above)
Comment (maflcko) [reply .github/actions/configure-docker/action.yml]: thx, took your doc comment

File: ci/test/02_run_container.py (hunk 13 above)
Comment (maflcko) [reply ci/test/02_run_container.py]: (removed)

File: ci/test/02_run_container.py (hunk 12 above)
Comment (maflcko) [reply ci/test/02_run_container.py]: Yes, exactly, see also the commit message :)

File: .github/actions/configure-docker/action.yml (hunk 11 above)
Comment (maflcko) [reply .github/actions/configure-docker/action.yml:52]: In theory yes, but in practise it does not work due to a docker bug. Also, I don't think it is worth it in practise, see https://github.com/bitcoin/bitcoin/pull/33620

File: ci/test/02_run_container.py (hunk 10 above)
Comment (maflcko) [reply ci/test/02_run_container.py]: (removed)

File: ci/test/02_run_container.py (hunk 9 above)
Comment (maflcko) [reply ci/test/02_run_container.py:67]: No, a follow-up requires it to be separate :)

File: ci/test/02_run_container.py (hunk 8 above)
Comment (maflcko) [reply ci/test/02_run_container.py]: Yes, `CI_IMAGE_NAME_TAG` is just a general `ubuntu:24.04`. I think logging the more specific name, like `ci_native_tidy` seems useful.

File: ci/test/02_run_container.py (hunk 7 above)
Comment (maflcko) [reply ci/test/02_run_container.py]: thx, inlined

File: ci/test/02_run_container.py (hunk 6 above)
Comment (maflcko) [reply ci/test/02_run_container.py]: (yes, it should be irrelevant, but removed)

File: ci/test/02_run_container.py (hunk 5 above)
Comment (maflcko) [reply ci/test/02_run_container.py]: (removed)

File: ci/test/02_run_container.py (hunk 4 above)
Comment (maflcko) [reply ci/test/02_run_container.py]: Yeah, it is a bit undocumented, and currently unused after the CI migration. Removed it for now. If there is ever need in the future to re-add it, it can be done properly. For future reference, the current state was:

```py
def maybe_cpuset():
    # Env vars during the build can not be changed. For example, a modified
    # $MAKEJOBS is ignored in the build process. Use --cpuset-cpus as an
    # approximation to respect $MAKEJOBS somewhat, if cpuset is available.
    if os.getenv("HAVE_CGROUP_CPUSET"):
        P = int(run(["nproc"], stdout=subprocess.PIPE).stdout)
        M = min(P, int(os.getenv("MAKEJOBS").lstrip("-j")))
        cpus = sorted(random.sample(range(P), M))
        return [f"--cpuset-cpus={','.join(map(str, cpus))}"]
    return []

File: ci/test/02_run_container.py (hunk 9 above)
Comment (l0rinc) [reply ci/test/02_run_container.py:67]: In the followup please consider avoiding concats (e.g. via `*list`), as shown above

**Review by l0rinc — APPROVED** [review]
Code review ACK fabe0e07de1ad2f26da62f3ebe0e9be3f939b1f8

Let's see how much slowdown this would cause without caching.

**Comment by maflcko** [comment]
> Let's see how much slowdown this would cause without caching.

I don't think this has any measurable effect on the runtime. The only benefit could be avoiding intermittent network errors.

**Comment by m3dwards** [comment]
ACK fabe0e07de1ad2f26da62f3ebe0e9be3f939b1f8

This will save a lot of space and help to prevent the other caches from getting evicted which are much more important than the docker build layer caches. Should only slow down the jobs by a few minutes at most and hopefully be a net positive as the other more important caches (like ccache) will not have been evicted.

Ran on master on my own fork.

**Review by m3dwards — APPROVED** [review]

**Comment by cedwies** [comment]
reACK fabe0e0

**Review by willcl-ark — APPROVED** [review]
ACK fabe0e07de1ad2f26da62f3ebe0e9be3f939b1f8

Agree with the rationale here and the changes look correct to me. There is a run on-going on a fork (non Cirrus runners) at https://github.com/m3dwards/bitcoin/actions/runs/18709746522

...and the [caches in that repo](https://github.com/m3dwards/bitcoin/actions/caches) do not contain docker build images as intended.

**Comment by maflcko** [comment]
Hmm, so I did a fresh run in a repo with a clean GHA cache, and it now says:

>  Approaching total cache storage limit (8.89 GB of 10 GB Used) 

So I guess we are still close to overlapping the cache and certainly will reach cache churn with more than one push to the default branch.
